Sensor Model's via Plugins

DanMcGann opened the pull request:
Draft of octomap sensor models using pluginlib plugins.

Definitely more overhead than the config only option in #11, but would make it much easier for adding points based individual return intensities if we plan to go that route in the future.



Example configuration 
```
isa_sensor_model:
        prob_hit: 0.58
        prob_miss: 0.4
        max_range: 40

gemini_sensor_model:
        prob_hit: 0.65
        prob_miss: 0.3
        max_range: 20

plugins: 
        - {name: isa_sensor_model,      type: "square_robot::Isa500ModelPlugin", frame_id: isa500_sensor}
        - {name: isa_sensor_model,      type: "square_robot::Isa500ModelPlugin", frame_id: isa500_sensor_2}
        - {name: gemini_sensor_model,   type: "square_robot::GeminiModelPlugin", frame_id: tritech_gemini_base}
```

Edit: Meant to make this a draft PR, and unfortunately there is no way to convert an existing PR to a Draft PR.

DanMcGann committed:
OctoMap Sensor Models using pluginlib.

DanMcGann committed:
Made header guards rooted in octomap_mapping.

DanMcGann committed:
Made use of Sensor Plugins optional

brenden-gibbons commented on octomap_server/include/octomap_server/OctomapServer.h:
```diff
@@ -285,6 +289,11 @@ class OctomapServer {
   bool m_useColoredMap;
+
+  // sensor models
+  bool m_use_sensor_plugins;
```
tiny nitpick about these member variable names.  Should be `m_usSensorPlugins`, `m_pluginLoader`, and `m_sensorModelMap` to match the preexisting style.

brenden-gibbons commented on sensor_model_plugins/plugins/isa500_model_plugin.cpp:
```diff
@@ -0,0 +1,31 @@
+PLUGINLIB_EXPORT_CLASS(square_robot::Isa500ModelPlugin, square_robot::SensorModelBase)
+
+namespace square_robot {
+Isa500ModelPlugin::Isa500ModelPlugin() : SensorModelBase(0.58, 0.4, 40) {
```
Even though we've never seen ISA ranges greater than 40 meters, we hope to one day!  Could you bump this to 60 to keep our dreams alive?

brenden-gibbons commented on octomap_server/src/OctomapServer.cpp:
```diff
@@ -493,13 +539,13 @@ void OctomapServer::insertScan(const tf::Point& sensorOriginTf, const PCLPointCl
   // now mark all occupied cells:
   for (KeySet::iterator it = occupied_cells.begin(), end=occupied_cells.end(); it!= end; it++) {
-    m_octree->updateNode(*it, true);
+    m_octree->updateNode(*it, logodds(occ_cell_probs.at(*it)), false);
```
I think this should be _m_octree->updateNode(*it, logodds(occ_cell_probs.at(*it)), **true**);_

sbiddlestone replied:
false is correct (technically it has a default value, so it could be removed like the previous one):
virtual NODE* updateNode(const OcTreeKey& key, float log_odds_update, bool lazy_eval = false);

The previous one was updateNode:
virtual NODE* updateNode(const OcTreeKey& key, bool occupied, bool lazy_eval = false);

sbiddlestone replied:
Method parameter overloading is fun!

brenden-gibbons reviewed (changes requested):
This is great! The pluginlib integration looks very clean.  My only real comment is about the occupied cells update updateNode call

DanMcGann committed:
Review fixes.

- Fixed naming conventions
- Removed default param from updateNode call
- bumped up the isa500's max range

DanMcGann committed:
Fixed map access bug when rays are out of range.

DanMcGann commented:
Caught a small but important bug in d23a803.
Also @brenden-square should we add a model plugin for the me1512 before merge?

DanMcGann committed:
Added sensor model for me1512
